### PR TITLE
Blocks E2E: Streamline E2E utils implementation

### LINKED ISSUE #47061: Blocks E2E: Update documentation
Let's update our [E2E docs](https://github.com/woocommerce/woocommerce/blob/8ee74ede0ebc1957089af08207ae2a1a8250610e/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md) to reflect the current state of things after completing the tasks from https://github.com/woocommerce/woocommerce/issues/46328.

### PR DESCRIPTION
**Pull request by WunderBart**
Closes #47061 

### What?
This PR aims to streamline the E2E utils implementation with the following changes:

- **Rebuild the file structure:** Combine `@woocommerce/e2e-playwright-utils`, `@woocommerce/e2e-types`, and `@woocommerce/e2e-utils` into a single `@woocommerce/e2e-utils` package. 
  - All utilities are now Playwright utilities, so it makes sense to keep them together. 
  - Following Gutenberg's E2E and Playwright conventions, types are exported along with the main package. Also, there was [just one shared type](https://github.com/woocommerce/woocommerce/blob/fa4f3ff3f787ff5062ba50185c6fa8c364e71191/plugins/woocommerce-blocks/tests/e2e/utils/types.ts) - `BlockData`. 😄 

- **Merge `EditorUtils` with Core's `Editor`:**
  - The purposes of both were identical, leading to duplicated functionality. Merging them reduces confusion and maintenance overhead.

- **Merge `TemplateAPIUtils` with `RequestUtils`:**
  - `TemplateAPIUtils` were just a set of REST calls, for which we generally use the `RequestUtils`. The [templates API calls subset](https://github.com/WordPress/gutenberg/blob/c7036160a6d488e660d9557512a9f4f2d1918004/packages/e2e-test-utils-playwright/src/request-utils/templates.ts) already exists in `RequestUtils`, so it was extended as needed.

- **Streamline Handlebars templating utils:** CC: @samueljseay @kmanijak
  - Aligned the implementation with how test plugins and themes are handled.
  - Created [a dedicated folder](https://github.com/woocommerce/woocommerce/blob/4b6e85a52ddd11e1b1b59e7c181968fa776b54a3/plugins/woocommerce-blocks/tests/e2e/content-templates) for Handlebars templates instead of keeping them alongside test files.
  - Defined the [`createPostFromFile`](https://github.com/woocommerce/woocommerce/blob/7df51cfb7f3108d61a3d76544ea9361027475af9/plugins/woocommerce-blocks/tests/e2e/utils/request-utils/posts.ts#L22) and [`createTemplateFromFile`](https://github.com/woocommerce/woocommerce/blob/892c90f361536465bd20d57c1223bceaa9cb4eb2/plugins/woocommerce-blocks/tests/e2e/utils/request-utils/templates.ts#L67) as `RequestUtils` extensions. For more details, please see [the updated guide on content templating](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md#content-templates).

- **Move one-off utils to respective suites to reduce over-abstraction:**
  - Many one-off utils, particularly in the `EditorUtils` class, have been moved to their respective test suites to avoid premature optimization and reduce maintenance complexity. Unless a utility is used extensively across many tests, it should be inlined or defined within the suite that uses it.

- **Remove unused utils:**
  - Unused utilities have been removed to clean up the codebase.

- **Update the [Blocks E2E readme](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md):**
  - All the changes above were reflected in the readme.

### How to test

- All tests should pass.
- Ensure [the new readme](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md) makes sense. 🙏

### CODE REVIEW

**Comment by WunderBart** [comment]
I'm not sure about the relation of the [contributors/e2e-guidelines.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md) to [tests/e2e/README.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md). I've updated only the latter so far. Any ideas on how to handle that? cc: @gigitux @Aljullu

File: plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md
Comment (nielslange) [inline plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md]: What are your thoughts on showing an example call, e.g. 

```js
import {
	test as base,
	expect,
	Editor,
	BlockData,
} from '@woocommerce/e2e-utils';
```

File: plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md
Comment (nielslange) [inline plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md]: For https://developer.woocommerce.com/docs/, we've created [Technical Documentation Style Guide](https://developer.woocommerce.com/docs/technical-documentation-style-guide/) a while ago. For the docs titles and subheadings, we have the following recomendation:
> Use sentence case (not title case) for docs titles and subheadings. Example: “Introduction to the launch experience” rather than “Introduction to the Launch Experience.”

While you haven't changed this particular title, what do you think about adjusting the titles on this doc to be in synch with our _Technical Documentation Style Guide_?

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (nielslange) [inline plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:128]: Is there a reason why we're not using the constant `SHOP_PAGE` from `@woocommerce/e2e-utils`?

**Comment by gigitux** [comment]
> I'm not sure about the relation of the [contributors/e2e-guidelines.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md) to [tests/e2e/README.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md). I've updated only the latter so far. Any ideas on how to handle that? cc: @gigitux @Aljullu

Good question! 🤔 I think that the contributor's folder contains all the best practices for writing JS/PHP/CSS code for contributors.

I think that we could move the `Writing Tests` section from [tests/e2e/README.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md) to [contributors/e2e-guidelines.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md).

I would love to hear @nielslange's feedback!

File: plugins/woocommerce-blocks/.wp-env.json
Comment (gigitux) [inline plugins/woocommerce-blocks/.wp-env.json:19]: Is there a particular reason for not using the `themes` option?

File: plugins/woocommerce-blocks/tests/e2e/tests/add-to-cart-form/add-to-cart-form.block_theme.spec.ts
Comment (gigitux) [inline plugins/woocommerce-blocks/tests/e2e/tests/add-to-cart-form/add-to-cart-form.block_theme.spec.ts:45]: NIT: We could potentially create a dedicated file for this class and call it `SingleProductBlockPage`.

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (nielslange) [inline plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:24]: I noticed that we're using this pattern in various files:

```ts
const test = base.extend< { templateCompiler: TemplateCompiler } >( {
	templateCompiler: async ( { requestUtils }, use ) => {
		const compiler = await requestUtils.createTemplateFromFile(
			'archive-product_attribute-filter'
		);
		await use( compiler );
	},
} );
```
```ts
const test = base.extend< { templateCompiler: TemplateCompiler } >( {
	templateCompiler: async ( { requestUtils }, use ) => {
		const compiler = await requestUtils.createTemplateFromFile(
			'archive-product_price-filter'
		);
		await use( compiler );
	},
} );
```
```ts
const test = base.extend< { templateCompiler: TemplateCompiler } >( {
	templateCompiler: async ( { requestUtils }, use ) => {
		const compiler = await requestUtils.createTemplateFromFile(
			'archive-product_rating-filter'
		);
		await use( compiler );
	},
} );
```

What do you think of creating a helper function for this, to prevent duplicated code in the individual test files?

File: plugins/woocommerce-blocks/tests/e2e/tests/on-sale-badge/on-sale-badge-single-product-template.block_theme.spec.ts
Comment (gigitux) [inline plugins/woocommerce-blocks/tests/e2e/tests/on-sale-badge/on-sale-badge-single-product-template.block_theme.spec.ts:73]: same here

File: plugins/woocommerce-blocks/tests/e2e/utils/types.ts
Comment (gigitux) [inline plugins/woocommerce-blocks/tests/e2e/utils/types.ts:6]: When I created this type, the idea was to "enforce" folks to keep the same style when they define the object for store selectors/slug, but this type isn't used so much. Do you think that we should keep it? I'm fine with removing it.

**Review by nielslange — COMMENTED** [review]
Thanks for working on this, @WunderBart. I left a few comments below.

**Review by gigitux — APPROVED** [review]
Brilliant work ✨! I'm pre-approving this PR! I left some comments, but mostly to hear your thoughts!  

Before merging, I would wait for @nielslange's feedback on the documentation!

**Comment by nielslange** [comment]
> I think that we could move the Writing Tests section from [tests/e2e/README.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md) to [contributors/e2e-guidelines.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md).

I like this suggestion, @gigitux. Regarding moving, I would only move the content, not the entire file. In `tests/e2e/README.md` I would then reference to `contributors/e2e-guidelines.md`. I did something similar in https://github.com/woocommerce/woocommerce-blocks/pull/11336/files, when moving docs to new locations.

cc: @WunderBart

**Comment by nielslange** [comment]
@WunderBart, I wanted to point out a lint error I mentioned in p1716786870965659/1716534287.104199-slack-C060B25RT41. I noticed you reacted to my comment, but for easier reference, I'm sharing the link to the internal discussion here.

**Comment by nielslange** [comment]
> Before merging, I would wait for @nielslange's feedback on the documentation!

I just replied in https://github.com/woocommerce/woocommerce/pull/47660#issuecomment-2132961794, and I agree that moving the e2e docs content is a great idea to make that information more findable.

File: plugins/woocommerce-blocks/tests/e2e/utils/types.ts
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/utils/types.ts:6]: IMO, it's currently used enough to justify a dedicated export (31 instances across 15 files). However, I agree that it's not very useful as it adds a layer of complexity to the tests and impairs the code readability. It's generally better to repeat a string a couple of times than to interpolate with a constant, e.g.:

```js
test.describe( `${ blockData.name } Block`, () => {
  test( 'Adding to cart', async ( { page } ) => {
    await page.locator( blockData.selectors.editor.button ).click();

    // ...
  } );
} );

// VS.

test.describe( 'Add to Cart Block', () => {
  test( 'Adding to cart', async ( { page } ) => {
    await page.getByLabel( "Add to Cart" ).click();

    // ...
  } );
} );
```

Therefore, I think we can drop the `BlockData` convention! 😄 I'm not sure, though, if this PR is a good place to do that, as it's already massive. Should we do it in a follow-up PR?

Comment (gigitux) [reply plugins/woocommerce-blocks/tests/e2e/utils/types.ts:6]: > Should we do it in a follow-up PR? 

Yeah, sure! 🙇

File: plugins/woocommerce-blocks/.wp-env.json
Comment (WunderBart) [reply plugins/woocommerce-blocks/.wp-env.json:19]: Not sure what you mean by 'themes option' - could you elaborate?

Comment (gigitux) [reply plugins/woocommerce-blocks/.wp-env.json:19]: You moved the themes URLs/paths from `themes` array to here:

image

Just wonder why :D

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:24]: I think this is the DRY-est version we should have at this point. 😅 There can be multiple compilers used for a single post, and for the opposite, one-off situations, we don't really need to extend the base like that. We could just do:

```js
test( 'My fancy post content', async ( { requestUtils } ) => {
  const myTemplate = await requestUtils.createTemplateFromFile( 'my-fancy-template' );
  await myTemplate.compile( { title: 'Howdy!' } );

  // ...
} );
```

Does that make sense?

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:128]: Hmm, I don't think we use that constant anywhere. 🤔 Having said that, this makes me realize that we have a conflict of aliases - our (WooCommerce Blocks) package has the same name as the [WC Core package](https://github.com/woocommerce/woocommerce/blob/ea25e2c5b2bbfa60626976fde5e5212c4038624b/packages/js/e2e-utils/README.md). Should we name our local utils `@woocommerce-blocks/e2e-utils` instead? cc @gigitux

File: plugins/woocommerce-blocks/tests/e2e/tests/add-to-cart-form/add-to-cart-form.block_theme.spec.ts
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/add-to-cart-form/add-to-cart-form.block_theme.spec.ts:45]: Yeah, I think we can think about creating a set of dedicated POMs in general. I think it's something for a larger discussion, though, so I'd leave it for a follow-up work, if that's OK.

File: plugins/woocommerce-blocks/.wp-env.json
Comment (WunderBart) [reply plugins/woocommerce-blocks/.wp-env.json:19]: Ah, gotcha! In this case it's better to just map the parent path, as it eliminates the need to define each theme's path explicitly. With the parent `theme` path mapped, you just need to throw the theme into the `e2e/themes` folder, and it's immediately picked up by the wp-env! This is also consistent with [how we configure it in Gutenberg](https://github.com/WordPress/gutenberg/blob/3f249471bf8bfed5cb0f9a7a15dd3b43bf3df464/.wp-env.json#L11).

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (gigitux) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:128]: Good catch! Let's change it, but I'm fine with doing it on a follow-up PR if you prefer! Given that currently our project is called `@woocommerce/block-library` not sure that `@woocommerce-blocks/e2e-utils` is a good name. 

What about something like `~/e2e-utils`? This pattern is used in `@woocommerce/admin-library` project: https://github.com/woocommerce/woocommerce/blob/7b753af893a374be213d88bbfee672196200f283/plugins/woocommerce-admin/tsconfig.json

Comment (nielslange) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:128]: > [...] Having said that, this makes me realize that we have a conflict of aliases - our (WooCommerce Blocks) package has the same name as the [WC Core package](https://github.com/woocommerce/woocommerce/blob/ea25e2c5b2bbfa60626976fde5e5212c4038624b/packages/js/e2e-utils/README.md).Having said that, this makes me realize that we have a conflict of aliases - our (WooCommerce Blocks) package has the same name as the [WC Core package](https://github.com/woocommerce/woocommerce/blob/ea25e2c5b2bbfa60626976fde5e5212c4038624b/packages/js/e2e-utils/README.md). [...]

Oh, I thought we were inheriting functionality from the core package and adding extra functionality to it. Are we using a complete separate package here?

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (nielslange) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:24]: Thanks for explaining this, Bart. Let's go with the current implementation, then.

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:128]: We're just [TS-aliasing our (Blocks) utilities](https://github.com/woocommerce/woocommerce/blob/fa4f3ff3f787ff5062ba50185c6fa8c364e71191/plugins/woocommerce-blocks/tsconfig.base.json#L70), which aren't related to the WC Core package. At some point in the future, though, I think having a single WooCommerce package would be the way to go. Pinging @lanej0 @adimoldovan @tammullen here for insight on potential work in this direction, i.e., the unification of the WooCommerce E2E framework, utilities, etc.

**Comment by WunderBart** [comment]
@gigitux @nielslange, I've split the general readme between the [tests/e2e/README.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md) and [docs/contributors/e2e-guidelines.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md) as suggested. Let me know if it makes sense to you!

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:128]: > What about something like ~/e2e-utils? This pattern is used in @woocommerce/admin-library project: https://github.com/woocommerce/woocommerce/blob/7b753af893a374be213d88bbfee672196200f283/plugins/woocommerce-admin/tsconfig.json

On second thought, let's hold off deciding on an alias here as [we currently seem to be aliasing](https://github.com/woocommerce/woocommerce/blob/fa4f3ff3f787ff5062ba50185c6fa8c364e71191/plugins/woocommerce-blocks/tsconfig.base.json#L27-L71) everything else inside the `woocommerce-blocks` plugin folder as `@woocommerce/*`.

**Comment by gigitux** [comment]
> @gigitux @nielslange, I've split the general readme between the [tests/e2e/README.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/tests/e2e/README.md) and [docs/contributors/e2e-guidelines.md](https://github.com/woocommerce/woocommerce/blob/refactor/core-e2e-utils-implementation/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md) as suggested. Let me know if it makes sense to you!

LGTM! 🚀

File: plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts
Comment (gigitux) [reply plugins/woocommerce-blocks/tests/e2e/tests/attributes-filter/attribute-filter.block_theme.spec.ts:128]: > On second thought, let's hold off deciding on an alias here as [we currently seem to be aliasing](https://github.com/woocommerce/woocommerce/blob/fa4f3ff3f787ff5062ba50185c6fa8c364e71191/plugins/woocommerce-blocks/tsconfig.base.json#L27-L71) everything else inside the woocommerce-blocks plugin folder as @woocommerce/*.


Yeah, I agree with you! It is a broader discussion! We didn't make any changes to this after the migration to the mono repo.

**Pull request merged**
